Core/include/Acts/Propagator/Propagator.ipp

```diff
@@ -27,80 +30,154 @@ template <typename S, typename N>
 template <typename propagator_state_t>
 auto Acts::Propagator<S, N>::propagate(propagator_state_t& state) const
     -> Result<void> {
-  // Pre-stepping call to the navigator and actor list
   ACTS_VERBOSE("Entering propagation.");
 
   state.stage = PropagatorStage::prePropagation;
 
-  // Pre-Stepping call to the actor list
+  // Pre-Propagation: call to the actor list, abort condition check
   state.options.actorList.act(state, m_stepper, m_navigator, logger());
-  // assume negative outcome, only set to true later if we actually have
-  // a positive outcome.
-
-  // start at true, if we don't begin the stepping loop we're fine.
-  bool terminatedNormally = true;
-
-  // Pre-Stepping: abort condition check
-  if (!state.options.actorList.checkAbort(state, m_stepper, m_navigator,
-                                          logger())) {
-    // Stepping loop
-    ACTS_VERBOSE("Starting stepping loop.");
-
-    terminatedNormally = false;  // priming error condition
-
-    // Propagation loop : stepping
-    for (; state.steps < state.options.maxSteps; ++state.steps) {
-      // Pre-Stepping: target setting
-      state.stage = PropagatorStage::preStep;
-      m_navigator.preStep(state, m_stepper);
-      // Perform a propagation step - it takes the propagation state
-      Result<double> res = m_stepper.step(state, m_navigator);
-      if (res.ok()) {
-        // Accumulate the path length
-        double s = *res;
-        state.pathLength += s;
-        ACTS_VERBOSE("Step with size = " << s << " performed");
-      } else {
-        ACTS_ERROR("Step failed with " << res.error() << ": "
-                                       << res.error().message());
-        // pass error to caller
-        return res.error();
+
+  if (state.options.actorList.checkAbort(state, m_stepper, m_navigator,
+                                         logger())) {
+    ACTS_VERBOSE("Propagation terminated without going into stepping loop.");
+
+    state.stage = PropagatorStage::postPropagation;
+
+    state.options.actorList.act(state, m_stepper, m_navigator, logger());
+
+    return Result<void>::success();
+  }
+
+  auto getNextTarget = [&]() -> Result<NavigationTarget> {
+    for (unsigned int i = 0; i < state.options.maxTargetSkipping; ++i) {
```

Comment:
I guess this is essentially a safety mechanism?

Reply:
Yes this is very similar to the maximum number of steps. A misconfigured or bugged navigator might never provide a reachable target and we would be stuck in a loop forever.